test/topology_custom/test_topology_upgrade_stuck.py

```diff
@@ -0,0 +1,148 @@
+#
+# Copyright (C) 2025-present ScyllaDB
+#
+# SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+#
+
+import asyncio
+import logging
+import pytest
+import time
+
+from test.pylib.manager_client import ManagerClient
+from test.pylib.rest_client import inject_error
+from test.pylib.util import wait_for_cql_and_get_hosts
+from test.topology.conftest import skip_mode
+from test.topology.util import reconnect_driver, enter_recovery_state, \
+        delete_raft_data_and_upgrade_state, log_run_time, wait_until_upgrade_finishes as wait_until_schema_upgrade_finishes, \
+        wait_until_topology_upgrade_finishes, delete_raft_topology_state, wait_for_cdc_generations_publishing, \
+        check_system_topology_and_cdc_generations_v3_consistency
+
+@pytest.mark.asyncio
+@skip_mode('release', 'error injections are not supported in release mode')
+@log_run_time
+async def test_topology_upgrade_stuck(request, manager: ManagerClient):
+    """
+    Simulates a situation where upgrade procedure gets stuck due to majority
+    loss: we have one upgraded node, one not upgraded node, and three nodes
+    permanently down. Then, it verifies that it's possible to perform recovery
+    procedure and redo the upgrade after the issue is resolved.
+    """
+
+    # First, force the first node to start in legacy mode
+    cfg = {'force_gossip_topology_changes': True, 'enable_tablets': False}
+
+    servers = [await manager.server_add(config=cfg) for _ in range(5)]
+    to_be_upgraded_node, to_be_isolated_node, *to_be_shutdown_nodes = servers
+    cql = manager.cql
+    assert(cql)
+
+    logging.info("Waiting until driver connects to every server")
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
```

Comment:
IIUC the "nice" way to do this nowadays is `get_ready_cql`

Reply:
Done